runtime/serde/serde-json/common/test/aws/smithy/kotlin/runtime/serde/json/JsonDeserializerTest.kt

```diff
@@ -553,4 +555,236 @@ class JsonDeserializerTest {
+                "string" to "foo"
+                "bool" to true
+                "null" to null as Int?
+            }
+        )
+
+    @Test
+    fun canDeserializeComplexDocument() {
+        val doc = """
+        {
+            "number": 12,
+            "string": "foo",
+            "bool": true,
+            "null": null,
+            "list": [
+                12.0,
+                "foo",
+                true,
+                null,
+                [
+                    12.0,
+                    "foo",
+                    true,
+                    null
+                ],
+                {
+                    "number": 12.0,
+                    "string": "foo",
+                    "bool": true,
+                    "null": null
+                }
+            ],
+            "map": {
+                "number": 12,
+                "string": "foo",
+                "bool": false,
+                "null": null,
+                "list": [
+                    12,
+                    "foo",
+                    false,
+                    null
+                ],
+                "map": {
+                    "number": 12,
+                    "string": "foo",
+                    "bool": false,
+                    "null": null
+                }
+            }
+        }
+        """
+            .replace("\n", "")
+            .replace(" ", "")
+        val expected = buildDocument {
+            "number" to 12L
+            "string" to "foo"
+            "bool" to true
+            "null" to null as Int?
+            "list" to buildList {
+                add(12.0)
+                add("foo")
+                add(true)
+                add(null as Int?)
+                add(
+                    buildList {
+                        add(12.0)
+                        add("foo")
+                        add(true)
+                        add(null as Int?)
+                    }
+                )
+                add(
+                    buildDocument {
+                        "number" to 12.0
+                        "string" to "foo"
+                        "bool" to true
+                        "null" to null as Int?
+                    }
+                )
+            }
+            "map" to buildDocument {
+                "number" to 12L
+                "string" to "foo"
+                "bool" to false
+                "null" to null as Int?
+                "list" to buildList {
+                    add(12L)
+                    add("foo")
+                    add(false)
+                    add(null as Int?)
+                }
+                "map" to buildDocument {
+                    "number" to 12L
+                    "string" to "foo"
+                    "bool" to false
+                    "null" to null as Int?
+                }
+            }
+        }
+
+        testDeserializeDocument(doc, expected)
+    }
+
+    private fun assertDocumentEquals(expected: Document, actual: Document) {
```

Comment:
I may be missing something here. Equality checks in Kotlin are normally value-based, not referential. That is, `assertEquals(listOf("foo"), listOf("foo"))` succeeds because `listOf("foo") == listOf("foo")`. Does this not work the same way for our document types which are `data class`es backed by data types which support value-based equality checking?

If this were strictly a testing concern I might be less insistent on understanding the answer. I'm concerned, however, with verifying our document type hierarchy supports idiomatic equality. For example, if a user has two copies of a document that are effectively the same, I'd want an equality check to verify that:

```kotlin
val docA = buildDocument {
    "foo" to 12
    "bar" to buildList { add(true); add("baz") }
}

val docB = buildDocument {
    "foo" to 12
    "bar" to buildList { add(true); add("baz") }
}

println(docA == docB) // Should print `true`
```

Reply:
Yeah, you're not wrong. At one point, tests with lists were failing and I was convinced it was because of this (like, it was literally saying "list a is not equal to list b" but the printed values were identical).

Stripping it out now and everything passes just fine, so it'll get removed.